File: algorithms/0015/main.go
package p0015

import (
	"sort"
)

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	result := [][]int{}
	for i := range nums {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		l, r := i+1, len(nums)-1
		for l < r {
			sum := nums[i] + nums[l] + nums[r]
			switch {
			case sum > 0:
				r--
			case sum < 0:
				l++
			default:
				result = append(result, []int{nums[i], nums[l], nums[r]})
				l, r = next(nums, l, r)
			}
		}
	}
	return result
}

func next(sort []int, l, r int) (int, int) {
	for l < r {
		switch {
		case sort[l] == sort[l+1]:
			l++
		case sort[r] == sort[r-1]:
			r--
		default:
			l++
			r--
			return l, r
		}
	}
	return l, r
}

File: algorithms/0015/main.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
  const sortNums = nums.sort((a, b) => a - b)
  const result = []

  for (let i = 0; i < sortNums.length; i++) {
    // 避免添加重复的结果
    if (i > 0 && sortNums[i] === sortNums[i - 1]) {
      continue
    }

    let l = i + 1
    let r = sortNums.length - 1
    while (l < r) {
      const sum = sortNums[i] + sortNums[l] + sortNums[r]
      if (sum > 0) {
        r -= 1
      } else if (sum < 0) {
        l += 1
      } else {
        result.push([sortNums[i], sortNums[l], sortNums[r]])
        // 计算下一次的 l, r
        const tmp = next(sortNums, l, r)
        l = tmp[0]
        r = tmp[1]
      }
    }
  }
  return result
}

function next(sort, l, r) {
  while (l < r) {
    if (sort[l] === sort[l + 1]) {
      l++
    } else if (sort[r] === sort[r - 1]) {
      r--
    } else {
      l += 1
      r -= 1
      return [l, r]
    }
  }
  return [l, r]
}

module.exports = threeSum
